File: runme/mcgen/local/datan/datsrc/auxzbr.c
#include "datsrc.h"

void LIBFUNC auxzbr_(double P_T *x0, double P_T *x1,
             double (CBFUNC P_T *funct)(double const P_T *,
             double const P_T *, integer const P_T *, integer const P_T *),
             double const P_T *par, integer const P_T *npar1,
             integer const P_T *npar2)
{
    integer i;
    double f0, f1, xs;

    if (*x0 == *x1) {
        *x1 = *x0 + 1.;
    }
    f0 = (*funct)(x0, par, npar1, npar2);
    f1 = (*funct)(x1, par, npar1, npar2);
    for (i = 1; i <= 1000; ++i) {
        if (f0 * f1 > 0.) {
            if (abs(f0) <= abs(f1)) {
                xs = *x0;
                *x0 += (*x0 - *x1) * 2.;
                *x1 = xs;
                f1 = f0;
                f0 = (*funct)(x0, par, npar1, npar2);
            } else {
                xs = *x1;
                *x1 += (*x1 - *x0) * 2.;
                *x0 = xs;
                f0 = f1;
                f1 = (*funct)(x1, par, npar1, npar2);
            }
        } else {
            goto L20;
        }
    }
L20:
    return;
} /* auxzbr_ */

File: runme/mcgen/local/datan/datsrc/auxzbr.f
      SUBROUTINE AUXZBR(X0,X1,FUNCT,PAR,NPAR1,NPAR2)
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      PARAMETER (ZERO=0.D0, ONE=1.D0, TWO=2.D0)
      EXTERNAL FUNCT
      IF(X0.EQ.X1) X1=X0+ONE
      F0=FUNCT(X0,PAR,NPAR1,NPAR2)
      F1=FUNCT(X1,PAR,NPAR1,NPAR2)
      DO 10 I=1,1000
        IF(F0*F1 .GT. ZERO) THEN
          IF(ABS(F0).LE.ABS(F1)) THEN
            XS=X0
            X0=X0+TWO*(X0-X1)
            X1=XS
            F1=F0
            F0=FUNCT(X0,PAR,NPAR1,NPAR2)
          ELSE
            XS=X1
            X1=X1+TWO*(X1-X0)
            X0=XS
            F0=F1
            F1=FUNCT(X1,PAR,NPAR1,NPAR2)
          END IF
        ELSE
          GO TO 20
        END IF
   10 CONTINUE
   20 CONTINUE
      END
